File: cometd-java/cometd-java-client/src/main/java/org/cometd/client/transport/TransportRegistry.java
/*
 * Copyright (c) 2008-2021 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.cometd.client.transport;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class TransportRegistry {
    private final Map<String, ClientTransport> _transports = new HashMap<String, ClientTransport>();
    private final List<String> _allowed = new ArrayList<String>();

    public void add(ClientTransport transport) {
        if (transport != null) {
            _transports.put(transport.getName(), transport);
            _allowed.add(transport.getName());
        }
    }

    public Set<String> getKnownTransports() {
        return Collections.unmodifiableSet(_transports.keySet());
    }

    public List<String> getAllowedTransports() {
        return Collections.unmodifiableList(_allowed);
    }

    public List<ClientTransport> negotiate(Object[] requestedTransports, String bayeuxVersion) {
        List<ClientTransport> list = new ArrayList<ClientTransport>();

        for (String transportName : _allowed) {
            for (Object requestedTransportName : requestedTransports) {
                if (requestedTransportName.equals(transportName)) {
                    ClientTransport transport = getTransport(transportName);
                    if (transport.accept(bayeuxVersion)) {
                        list.add(transport);
                    }
                }
            }
        }
        return list;
    }

    public ClientTransport getTransport(String transport) {
        return _transports.get(transport);
    }
}

File: cometd-javascript/cometd-javascript-common/src/main/webapp/js/cometd/TransportRegistry.js
/*
 * Copyright (c) 2008 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * A registry for transports used by the CometD object.
 */
export class TransportRegistry {
    #types = [];
    #transports = {};

    getTransportTypes() {
        return this.#types.slice(0);
    }

    add(type, transport, index) {
        let existing = false;
        for (let i = 0; i < this.#types.length; ++i) {
            if (this.#types[i] === type) {
                existing = true;
                break;
            }
        }

        if (!existing) {
            if (typeof index !== "number") {
                this.#types.push(type);
            } else {
                this.#types.splice(index, 0, type);
            }
            this.#transports[type] = transport;
        }

        return !existing;
    }

    find(type) {
        for (let i = 0; i < this.#types.length; ++i) {
            if (this.#types[i] === type) {
                return this.#transports[type];
            }
        }
        return null;
    }

    negotiateTransport(types, version, crossDomain, url) {
        for (let i = 0; i < this.#types.length; ++i) {
            const type = this.#types[i];
            for (let j = 0; j < types.length; ++j) {
                if (type === types[j]) {
                    const transport = this.#transports[type];
                    if (transport.accept(version, crossDomain, url) === true) {
                        return transport;
                    }
                }
            }
        }
        return null;
    }

    clear() {
        this.#types = [];
        this.#transports = {};
    }

    reset(init) {
        for (let i = 0; i < this.#types.length; ++i) {
            this.#transports[this.#types[i]].reset(init);
        }
    }

    findTransportTypes(version, crossDomain, url) {
        const result = [];
        for (let i = 0; i < this.#types.length; ++i) {
            const type = this.#types[i];
            if (this.#transports[type].accept(version, crossDomain, url) === true) {
                result.push(type);
            }
        }
        return result;
    }

    remove(type) {
        for (let i = 0; i < this.#types.length; ++i) {
            if (this.#types[i] === type) {
                this.#types.splice(i, 1);
                const transport = this.#transports[type];
                delete this.#transports[type];
                return transport;
            }
        }
        return null;
    }
}
